### ISSUE TITLE
[Workload] EPIC: Integrate test cases for kyverno integration

### ISSUE THREAD

**Issue by lixuna** [cnti-testcatalog/testsuite#1182]
**Is your workload test idea related to a problem? Please describe.**
- This issue is for tracking the iterative reviews and merging tests included in PR #1176 

**Describe the solution you'd like**
- The PR will be broken into smaller PRs by the reviewers and merged iteratively

---

### Documentation QA:
- [ ] Update [installation instructions](https://github.com/cncf/cnf-testsuite/blob/main/install.md) if needed
- [ ] Update [Test Categories md](https://github.com/cncf/cnf-testsuite/blob/main/TEST-CATEGORIES.md) if needed
- [ ] Update [USAGE md](https://github.com/cncf/cnf-testsuite/blob/main/USAGE.md) if needed
  - [ ] How to run
  - [ ] Description and details
    - [ ] What the best practice is
    - [ ] Why are we testing this
  - [ ] Remediation steps if test does not pass

### QA tasks

Dev Review:

- [ ] walk through A/C
- [ ] do you get the expected result?
- [ ] if yes,
  - [ ] move to `Needs Peer Review` column
  - [ ] create Pull Request and follow check list
  - [ ] Assign 1 or more people for peer review
- [ ] if no, document what additional tasks will be needed

Peer review:

- [ ] walk through A/C
- [ ] do you get the expected result?
- [ ] if yes,
  - [ ] move to `Reviewer Approved` column
  - [ ] Approve pull request
- [ ] if no,
  - [ ] document what did not go as expected, including error messages and screenshots (if possible)
  - [ ] Add comment to pull request
  - [ ] request changes to pull request

**Comment by HashNuke** [comment]
Made some notes on tests in the PR to begin splitting into separate branches.

<img width="554" alt="image" src="https://user-images.githubusercontent.com/84005/156553074-9d01c36c-1754-43c2-9b18-f6e2a32c8877.png">

**Comment by agentpoyo** [comment]
`restrict_external_ips` was also already merged into `main`. @HashNuke

**Comment by agentpoyo** [comment]
We still need spec tests and sample cnf if required done for https://github.com/cncf/cnf-testsuite/issues/1233 https://github.com/cncf/cnf-testsuite/issues/1234 https://github.com/cncf/cnf-testsuite/issues/1235 and https://github.com/cncf/cnf-testsuite/issues/1236

**Comment by HashNuke** [comment]
@agentpoyo Ah yes. My bad. I missed marking that properly in the sheet above.

1. I'm beginning with adding spec tests for the 4 tests already in main.
2. I'll integrate the other tests once these above is done.

**Comment by HashNuke** [comment]
**[Update]**
Current output message structure is in the below screenshot. When making updates to container_sock_mounts test, I came across better message in the actual PolicyReport that can be used as remediation message.

<img width="965" alt="image" src="https://user-images.githubusercontent.com/84005/157008711-df66d959-5c65-453c-923f-59f2a74bea61.png">


It would be better if we use the messages in the PolicyReport since they are more specific. For example, the two validation issues in the screenshot are for the same resource and for the same test. But there are two outputs because one of the volume mount is a docker engine socket. The other volume mount is a containerd engine socket.

**Comment by HashNuke** [comment]
[Update] Regarding the `disallow_helm_tiller` test
* Renamed `disallow_helm_tiller` to `helm_tiller`
* Moved the test from workload security to platform security (`platform:helm_tiller`)

**Comment by HashNuke** [comment]
Tests completed in the `kyverno-improvements` branch:
* `container_sock_mounts` (previously called `disallow_container_sock_mounts`)
* `require_labels`
* `helm_tiller` (previously called `disallow_helm_tiller`)
* `external_ips` (previously called `restrict_external_ips`)

**Comment by HashNuke** [comment]
Added a comment with notes about changing Kyverno to a cli-based utility within the testsuite on another ticket (#1247).

Updated the PR with appropriate changes too (#1266).

**Comment by HashNuke** [comment]
## Tests that require discussion

#### `restrict_image_registries` ([Kyverno policy file](https://raw.githubusercontent.com/kyverno/policies/main/best-practices/restrict_image_registries/restrict_image_registries.yaml))
* The policy file is a sample that has to be updated as per requirement.
* AFAIK this is not something that we can standardise as a guideline. In order to implement this test, the testsuite would have to allow the user to configure registries that can be used.

#### `restrict_volume_types` ([Kyverno policy file](https://raw.githubusercontent.com/kyverno/policies/main/pod-security/restricted/restrict-volume-types/restrict-volume-types.yaml))
* AFAIK this is not something that we can standardise as a guideline. This test would require that the testsuite or the user configure allowed volume types that are allowed.
* Alternative: If we can identify certain volume types that the CNF should not be using, then we can use this test to check if the volume types are not used.

#### `add_network_policy` ([Kyverno Policy file](https://github.com/kyverno/policies/blob/main/best-practices/add_network_policy.yaml))
* The policy file "adds" a default network policy for namespaces. It does not test for anything.
* Created a ticket for `add_network_policy` before actually getting to review the policy file - #1217

## Overlapping tests

I reviewed the Kyverno policy files for all tests added in PR #1176. Here are tests in the PR that overlap with existing tests in the testsuite.

1. `deny_privilege_escalation`
    * [Kyverno Policy file](https://github.com/kyverno/policies/blob/main/pod-security/restricted/disallow-privilege-escalation/disallow-privilege-escalation.yaml)
    * This is the same as the kubescape-based test `privilege_escalation` 
2. `disallow_add_capabilities`
    * [Kyverno Policy file](https://github.com/kyverno/policies/blob/main/psp-migration/restrict_adding_capabilities/restrict-adding-capabilities.yaml)
    * Overlaps with two Kubescape-based tests `dangerous_capabilities` and `insecure_capabilities`
3. `disallow_host_namespaces`
    * [Kyverno Policy file](https://github.com/kyverno/policies/blob/main/pod-security/baseline/disallow-host-namespaces/disallow-host-namespaces.yaml)
    * Overlaps with two kubescape-based tests `host_network` and `host_pid_ipc_privileges`
4. `disallow_host_path`
    * [Kyverno policy file](https://raw.githubusercontent.com/kyverno/policies/main/pod-security/baseline/disallow-host-path/disallow-host-path.yaml)
    * Overlaps with Kubescape-based `hostpath_mounts` test.
5. `disallow_host_ports`
    * [Kyverno Policy file](https://github.com/kyverno/policies/blob/main/pod-security/baseline/disallow-host-ports/disallow-host-ports.yaml)
    * Overlaps with `hostport_not_used` test.
6. `disallow_privileged_containers `
    * [Kyverno Policy file](https://raw.githubusercontent.com/kyverno/policies/main/pod-security/baseline/disallow-privileged-containers/disallow-privileged-containers.yaml)
    * Overlaps with `privileged_containers` test.
7. `require_non_root_groups`
    * The Kyverno policy file url used in the PR does not exist in the repo anymore. [This might be the new Kyverno policy file](https://github.com/kyverno/policies/blob/main/pod-security/restricted/require-run-as-non-root-user/require-run-as-non-root-user.yaml)
    * Overlaps with existing `non_root_containers` test.
8. `require_requests_limits`
    * [Kyverno policy file](https://raw.githubusercontent.com/kyverno/policies/main/best-practices/require_pod_requests_limits/require_pod_requests_limits.yaml)
    * Overlaps with existing Kubescape-based `resource_policies` test.
9. `require_ro_rootfs`
    * [Kyverno Policy file](https://raw.githubusercontent.com/kyverno/policies/main/best-practices/require_ro_rootfs/require_ro_rootfs.yaml)
    * Overlaps with `immutable_file_systems` test.
10. `restrict_node_port`
    * [Kyverno policy file](https://raw.githubusercontent.com/kyverno/policies/main/best-practices/restrict_node_port/restrict_node_port.yaml)
    * Overlaps with `nodeport_not_used` test.
11. `run_as_non_root`
    * [Kyverno policy file](https://raw.githubusercontent.com/kyverno/policies/main/pod-security/restricted/require-run-as-nonroot/require-run-as-nonroot.yaml)
    * Overlaps with `non_root_containers` test.

**Comment by lixuna** [comment]
Hi @HashNuke Can this issue be closed?
